rfd/00xx-pod-access-requests.md

```diff
@@ -0,0 +1,598 @@
+the Teleport RBAC layer can extract the `{name}` from the request
+URI and validate if the pod name is in the user's allowed list.
+
+Each time a user requests an endpoint prefixed with `/api/v1/namespaces/{namespace}/pods/{name}`,
+Teleport Kubernetes Service intercepts the request and extracts the `{name}`.
+After, it will validate if the `{name}` is included in the `kubernetes_pods`
+array or if it satisfies any wildcard pattern defined on it. If allowed, Teleport
+will forward the request to Kubernetes with the RBAC principals as Impersonation
+headers. Kubernetes will apply a second authorization layer using Kubernetes RBAC
+to validate the request. Finally, Teleport would forward the Kubernetes API
+response directly to the user without any modification.
+
+#### Special handlers
+
+When the Pod name is not accessible at request time, Teleport will have
+different behavior. Depending on the action the user is doing, Teleport will
+handle them separately.
+
+##### Listing Pods
+
+When listing Pods in the cluster or at a certain namespace, Teleport will receive
+the list of Pods from Kubernetes itself. Since Teleport depends on what Kubernetes
+RBAC allows users to access, Teleport would need to hijack the Kubernetes API
+response to exclude any Pod that is not included in the allowed list.
+
+To do that, Teleport will automatically forward the request to Kubernetes
+API for the following two endpoints:
+
+- `/api/v1/pods`
+  - `GET`: list or watch objects of kind Pods in every namespace.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `GET`: list or watch objects of kind Pods in namespace `{namespace}`.
+
+Once Teleport receives the response from Kubernetes API, it will unmarshal the
+response and exclude any Pod that is not in the allowed list. After excluding them,
+Teleport will return the filtered response to the user.
+
+##### Deleting Pods
+
+Kubernetes allows a user to delete every Pod in a certain namespace using a single
+call. The call does not include the name of the Pods to delete, so Teleport won't
+forward the request to Kubernetes. Instead, Teleport will list the user's Pods,
+exclude the ones that he shouldn't have access to, and issue a delete request for
+each one of the resulting Pod names.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `DELETE`: delete a collection of Pods in namespace `{namespace}`.
+
+Another option is to prevent this type of request in Teleport. This means that
+if you try to delete every Pod in the namespace while your role limits the Pods
+you have access to, Teleport will reject the request and inform the user that
+he is not allowed to do it. This option will restrict one of the `kubectl`
+functionalities.
+
+##### Creating Pods
+
+The endpoint for creating pods won't be restricted, and if restrictions exist
+it's the Kubernetes RBAC layer that applies them.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `POST`: create a POD in namespace `{namespace}`.
+
+### Security
```

Comment:
Teleport extracts the `kubernetes_groups/users` from the certificate and sends them as Impersonation headers to the target cluster. The cluster evaluates two things:
- if the auth credentials used have permission to impersonate users and groups. In this case, if teleport credentials allow Impersonation.
- if the Impersonation users and groups' RBAC policies allow the operation.

If teleport allows the resource, but the RBAC denies the request i.e. a user is allowed to access logs for pod X but tries to exec into that Pod, Teleport will let the request through and Kubernetes RBAC will block the request since none of the forwarded impersonation groups and users allow exec.

Once you join a pod, you can directly hit the cluster API but it uses the pod's service account credentials. In the scenario you described, the pod's service account permissions must allow explicitly access to other pods. By default, SA has no permissions. 
If the user tries to impersonate another kube user/group such as `system:masters`, the pod's SA RBAC must explicitly allow impersonation actions. If it doesn't, the request is denied.

The scenario you described depends on what the target pod permissions allow.